### PR TITLE
RavenDB-21174: Improvements to XXHash codegen.

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21174

### Additional description
Includes changes to the code to ensure the JIT can emit better code. 

### Type of change
- Optimization

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### Is it platform specific issue?
- No

### CODE REVIEW

**Comment by redknightlois** [comment]
30% improvement on average.
image

File: test/FastTests/Sparrow/Hashing.cs [hunk 1]
```diff
@@ -387,114 +284,250 @@ public unsafe void XXHash32_StreamedHashingEquivalence(int bufferSize)
...
 
-            int blockSize;
-            int iteration = 1;
-            do
+                    Assert.Equal(expected, result);
+                }
+            }
+        }
+
+        private static unsafe ulong XXHash64Reference(byte* buffer, ulong len, ulong seed = 0)
```
Comment (arekpalinski) [inline test/FastTests/Sparrow/Hashing.cs:317]: This is the old implementation, right?

File: test/FastTests/Sparrow/Hashing.cs [hunk 2]
```diff
@@ -315,54 +256,11 @@ public static IEnumerable<object[]> BufferSize
                     new object[] {128},
                     new object[] {129},
                     new object[] {1000},
+                    new object[] {Random.Shared.Next(10000)}
```
Comment (arekpalinski) [inline test/FastTests/Sparrow/Hashing.cs]: I think we won't see the number returned here once a test will fail. We have `[InlineDataWithRandomSeed]` for that purpose

File: test/FastTests/Sparrow/Hashing.cs [hunk 3]
```diff
@@ -387,114 +284,250 @@ public unsafe void XXHash32_StreamedHashingEquivalence(int bufferSize)
...
             while (blockSize <= bufferSize);
         }
 
-        [Theory]
-        [MemberData("BufferSize")]
-        public unsafe void XXHash64_StreamedHashingEquivalence(int bufferSize)
+        [Fact]
+        public unsafe void XXHash64_HashingEquivalenceWithReference()
         {
             var rnd = new Random(1000);
```
Comment (arekpalinski) [inline test/FastTests/Sparrow/Hashing.cs]: Can we pass a seed here? You can get it from [InlineDataWithRandomSeed]

File: test/FastTests/Sparrow/Hashing.cs [hunk 4]
```diff
@@ -387,114 +284,250 @@ public unsafe void XXHash32_StreamedHashingEquivalence(int bufferSize)
...
-        [MemberData("BufferSize")]
-        public unsafe void Metro128_StreamedHashingEquivalence(int bufferSize)
+
+        [Fact]
+        public unsafe void XXHash32_HashingEquivalenceWithReference()
         {
             var rnd = new Random(1000);
 
-            byte[] values = new byte[bufferSize];
+            byte[] values = new byte[1024];
```
Comment (arekpalinski) [inline test/FastTests/Sparrow/Hashing.cs]: Can have random sizes here?

**Comment by arekpalinski** [comment]
You need to adjust the assertion in the test https://jenkins.hibernatingrhinos.com/job/PR_Tests/8989/testReport/SlowTests.Tests/TestsInheritanceTests/AllTestsShouldUseRavenFactOrRavenTheoryAttributes/

because you removed some tests. In the future please use `[RavenFact] / [RavenTheory]` for new tests.

File: test/FastTests/Sparrow/Hashing.cs (hunk 1 above)
Comment (redknightlois) [reply test/FastTests/Sparrow/Hashing.cs:317]: Yes, this is the old implementation we use as reference.

File: test/FastTests/Sparrow/Hashing.cs (hunk 2 above)
Comment (redknightlois) [reply test/FastTests/Sparrow/Hashing.cs]: it should yes, but I like the usage of `[InlineDataWithRandomSeed]` more.

File: test/FastTests/Sparrow/Hashing.cs (hunk 4 above)
Comment (redknightlois) [reply test/FastTests/Sparrow/Hashing.cs]: Yes we can... will do so. Good call.

File: test/FastTests/Sparrow/Hashing.cs (hunk 3 above)
Comment (redknightlois) [reply test/FastTests/Sparrow/Hashing.cs]: Done in all of them.

**Comment by redknightlois** [comment]
Updated all tests, introduced a new category as none of the current were appropriate (`Core`) which is only for low level functionality.

File: test/FastTests/Sparrow/Hashing.cs [hunk 5]
```diff
@@ -286,215 +234,286 @@ public unsafe void EnsureZeroLengthStringIsAValidHash()
...
-        [MemberData("BufferSize")]
-        public unsafe void XXHash32_StreamedHashingEquivalence(int bufferSize)
+        [RavenTheory(RavenTestCategory.Core)]
+        [InlineDataWithRandomSeed]
+        public unsafe void XXHash64_StreamedHashingEquivalence(int randomSeed)
         {
-            var rnd = new Random(1000);
+            var rnd = new Random(randomSeed);
+            var bufferSize = rnd.Next(1, 10000);
+            var seed = (ulong)rnd.Next(1, 10000);
```
Comment (arekpalinski) [inline test/FastTests/Sparrow/Hashing.cs]: Do we need to limit it to be in a range (1, 10000)?

**Comment by arekpalinski** [comment]
> You need to adjust the assertion in the test https://jenkins.hibernatingrhinos.com/job/PR_Tests/8989/testReport/SlowTests.Tests/TestsInheritanceTests/AllTestsShouldUseRavenFactOrRavenTheoryAttributes/
> 
> because you removed some tests. In the future please use `[RavenFact] / [RavenTheory]` for new tests.

Reminder about adjusting this test. You need to change the number hard coded here:
https://github.com/ravendb/ravendb/blob/6070c7e148bf3067d418fa7900b9556f3d84d20e/test/SlowTests/Tests/TestsInheritanceTests.cs#L94

**Review by ayende — APPROVED** [review]

**Review by maciejaszyk — APPROVED** [review]

File: test/FastTests/Sparrow/Hashing.cs (hunk 5 above)
Comment (redknightlois) [reply test/FastTests/Sparrow/Hashing.cs]: We don't want the buffer too be too big and from the code itself there is no much difference after you hit the bulk loop. So we dont win anything, and we are limiting the probability of cases where it is important which is lower than 100. In fact, if you push me a little bit I would diminish that into 1000 instead.

**Comment by redknightlois** [comment]
Mhhh, what is that `numberToTolerate` ?

EDIT: Ohh, we are checking how many tests dont have the new attributes... got it. Updated

File: test/FastTests/Sparrow/Hashing.cs (hunk 5 above)
Comment (redknightlois) [reply test/FastTests/Sparrow/Hashing.cs]: I went ahead and shrink it a bit and allowed the seed to be whatever `ulong/uint`

**Review by ppekrol — APPROVED** [review]

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**